design/construct-tree.md

```diff
@@ -0,0 +1,343 @@
+        2. ***Properties*** - Constructs will opt into the set of properties, and these will be exposed. Higher level constructs (L2 and L3) will expose properties and values of the resources they contain and not necessarily enumerate every property that’s available for a resource. This will especially be useful for constructs that are opinionated and set defaults on behalf of developers. Low level constructs (L1) will expose all properties that are configured.
+        3. ***Errors/Warnings*** - Errors and warnings that are produced during synthesis indicating validation failures, deprecation notices, guidance, etc will be included in the tree. These are emitted at the construct level although the message may point towards a specific property. The construct tree should be complete and include validation failures for all resources in the CDK application rather than just just fail on the first error.
+    4. **Assets** - Assets information will be included in the construct tree. Assets represent actions that the CDK *will* take ahead of stack deployments (i.e. S3 assets are zipped and uploaded directly, Docker images are uploaded to ECR). The asset metadata in the tree is the S3 key that an asset *would* have
+3. **Local** - The construct tree is produced at synthesis time. Connectivity to AWS and CloudFormation should not be required to produce a tree view of a CDK application.
+
+## Approach
+
+At a high-level, we will render the `construct tree` through a construct tree data model and construct metadata interface that constructs will implement. This section will describe the components of the JSON structure that will represent the construct-tree:
+
+### Construct Tree Data Model
+
+A feature will be added to the CDK core module so that any CDK application will produce an artifact called `tree.json` during calls to `cdk synth` and include it as an output to the cloud assembly in `cdk.out`. Developers will be able to opt-out if this functionality is not desired or needed through a flag in the CLI or set it as a preference.
+
+The construct tree will be a list of paths that are indexed into a flat map of constructs. The file structure itself represents the tree. This information will need to include the following to start rendering a flat map of constructs:
+
+### Construct properties
+
+|Property    |Type    |Required |Description  |
+|---  |---  |---  |---  |
+|id |string |Required |id of the construct within the current scope |
+|path |string |Required |full, absolute path of the construct within the tree |
+|children |Array<Path>  |Not Required |All direct children of this construct. Array of the absolute paths of the constructs. Will be used to walk entire list of constructs |
+|metadata |Array<Metadata>  |Not Required |Metadata describing all constructs/resources that are encapsulated by the construct |
```

Comment:
Similar to the comment I had in the Requirements section.

What constitutes a 'metadata'? Are we using this word for a grab bag of things? Any grouping here needs to answer the question - "Is this new piece of information (xxx) I want to add to the JSON file of type 'metadata'?"

Why is `id` not metadata but `description`? Similarly, why is `path` not metadata?

From the top of my head, one suggestion I have is to have two groups - 'ConstructProperties' and 'SourceProperties' (or something along those lines). The former contains all information that the construct defines and exposes, such as, id, description, type, hierarchy, etc. The latter contains information that are derived from the source, such as, errors, warnings, sourcelocation, etc.